lib/rucio/daemons/transmogrifier/transmogrifier.py

```diff
@@ -188,269 +221,463 @@ def transmogrifier(bulk=5, once=False, sleep_time=60):
+        partition_wait_time=1,
+        sleep_time=sleep_time,
+        run_once_fnc=functools.partial(
+            run_once,
+            bulk=bulk,
+        ),
+    )
+
+
+def run_once(heartbeat_handler, bulk, **_kwargs):
+
+    worker_number, total_workers, logger = heartbeat_handler.live()
+    dids, subscriptions = [], []
+    tottime = 0
+    try:
+        #  Get the new DIDs based on the is_new flag
+        logger(logging.DEBUG, "Listing new dids")
+        for did in list_new_dids(
+            thread=worker_number,
+            total_threads=total_workers,
+            chunk_size=bulk,
+            did_type=None,
+        ):
+            dids.append(
+                {
+                    "scope": did["scope"],
+                    "did_type": str(did["did_type"]),
+                    "name": did["name"],
+                }
+            )
+        logger(logging.INFO, "%i new dids to process", len(dids))
+
+        sub_dict = {3: []}
+        #  Get the list of subscriptions. The default priority of the subscription is 3. 0 is the highest priority, 5 the lowest
+        #  The priority is defined as 'policyid'
+        logger(logging.DEBUG, "Listing active subscriptions")
+        for sub in list_subscriptions(None, None):
+            if (
+                sub["state"] != SubscriptionState.INACTIVE
+                and sub["lifetime"]
+                and (datetime.now() > sub["lifetime"])
+            ):
+                update_subscription(
+                    name=sub["name"],
+                    account=sub["account"],
+                    metadata={"state": SubscriptionState.INACTIVE},
+                )
+
+            elif sub["state"] in [SubscriptionState.ACTIVE, SubscriptionState.UPDATED]:
+                priority = 3
+                if "policyid" in sub:
+                    if int(sub["policyid"]) not in sub_dict:
+                        sub_dict[int(sub["policyid"])] = []
+                    priority = int(sub["policyid"])
+                sub_dict[priority].append(sub)
+        priorities = list(sub_dict.keys())
+        priorities.sort()
+        #  Order the subscriptions according to their priority
+        for priority in priorities:
+            subscriptions.extend(sub_dict[priority])
+        logger(logging.INFO, "%i active subscriptions", len(subscriptions))
+    except SubscriptionNotFound as error:
+        logger(logging.WARNING, "No subscriptions defined: %s" % (str(error)))
+        must_sleep = True
+        return must_sleep
+    except Exception as error:
+        logger(
+            logging.ERROR,
+            "Failed to get list of new DIDs or subscriptions: %s" % (str(error)),
+        )
+        must_sleep = False
+        return must_sleep
+
+    results = {}
+    start_time = time.time()
+    blocklisted_rse_id = [rse["id"] for rse in list_rses({"availability_write": False})]
+    identifiers = []
+    #  Loop over all the new dids
+    for did in dids:
```

Comment:
How long can it take to iterate over all dids ? maybe add heartbeat_handler.live() inside the loop ? Hearbeats have shorter

Reply:
I'm not sure it is necessary and even if it's a good idea.
Usually, the loop is pretty fast, e.g. :
```
2022-05-23 08:13:02,001 root    1       INFO    transmogrifier[3/13] : It took 1.240969 seconds to process 87 DIDs
2022-05-23 08:13:06,377 root    1       INFO    transmogrifier[4/13] : It took 6.169517 seconds to process 87 DIDs
2022-05-23 08:14:02,394 root    1       INFO    transmogrifier[4/13] : It took 0.878970 seconds to process 156 DIDs
2022-05-23 08:14:02,981 root    1       INFO    transmogrifier[3/13] : It took 1.352124 seconds to process 181 DIDs
2022-05-23 08:14:02,984 root    1       INFO    transmogrifier[2/13] : It took 1.619641 seconds to process 181 DIDs
2022-05-23 08:14:07,596 root    1       INFO    transmogrifier[1/13] : It took 7.792062 seconds to process 170 DIDs
2022-05-23 08:15:01,006 root    1       INFO    transmogrifier[1/13] : It took 0.445081 seconds to process 174 DIDs
2022-05-23 08:15:01,737 root    1       INFO    transmogrifier[4/13] : It took 0.657582 seconds to process 183 DIDs
2022-05-23 08:15:02,201 root    1       INFO    transmogrifier[2/13] : It took 1.030301 seconds to process 167 DIDs
2022-05-23 08:15:06,871 root    1       INFO    transmogrifier[3/13] : It took 5.895660 seconds to process 175 DIDs
```
If we add a hearbeat live for each did, it will increase by 2 orders of magnitude. IMHO, it's a bit too much.

Reply:
With the heartbeat_handler it should be fine. This will only fire off an actual DB interaction if the last heartbeat is quite some time ago. So you can actually call this method quite often without any impact.